### PR TITLE
fix: Rewrite the insert_before method on Terminal to fix a bug and remove the height cap

### PR DESCRIPTION
**Pull request by danny-burrows**
This PR fixes some issues encountered in #584 by way of rewriting the `insert_before` method.

> **The function signature for reference:** 
> ```rs
> pub fn insert_before<F>(&mut self, height: u16, draw_fn: F) -> io::Result<()>
> ```

Ultimately it does two things:

1. It fixes a bug triggered by the `height` parameter of a call to `insert_before` being greater than `self.viewport_area.height` - This bug would cause an extra newline to be inserted leaving a gap in the output.
2. Removes the `height` cap by rendering output in chunks. if `height` is too large to fit on the screen (with the viewport) then chunks are rendered as the screen is scrolled.

## A Note About Test Coverage
~~There are currently no tests for the `insert_before` method; I've attempted to test this change with the `TestBackend` but unfortunately due to it not supporting clear `ClearType::AfterCursor` this would prove difficult and I think the noise of adding this functionality wouldn't be worth adding to this PR. As it's a very minimal hit I wondered if it's acceptable?~~

> **Update:** I've now added some test cases for the issues encountered in the `tests/` folder. I know this is now discouraged but there was already a pattern of using the `TestBackend` in this location, but I'd be open to moving them if requested. I've also added the `clear_region` and `append_lines` methods to the `TestBackend` with accompanying tests.

### CODE REVIEW

**Comment by joshka** [comment]
>There are currently no tests for the insert_before method; I've attempted to test this change with the TestBackend but unfortunately due to it not supporting clear ClearType::AfterCursor this would prove difficult and I think the noise of adding this functionality wouldn't be worth adding to this PR. As it's a very minimal hit I wondered if it's acceptable?

I think it's probably worth adding the clear ops to the TestBackend to help with this change. There are three main reasons I like to add test coverage to existing code before changing it:

1. When the existing code is complex or just not well understood, tests help document the existing behavior (particularly the edge cases)
2. When the existing code is changed, the tests help document how the behavior changes from the perspective of the consuming code
3. When the existing code is tested, any new implementation can be fearlessly refactored for simplicity and readability while maintaining correctness of implementation

These reasons each make changes easier to review as otherwise a reviewer has to determine whether the code did the right thing originally by fully understanding the code, and whether it now still does the right thing by understanding the new code.

My approach generally would be to write characterization tests that fully cover the problem code, commit them, then make the changes and see which tests break. How about adding the missing TestBackend implementations for clear that would show how this works and then adding the tests for this? Do you think this sound reasonable?

**Comment by danny-burrows** [comment]
That's reasonable yes, I'll work on adding that functionality to the `TestBackend` and some tests to showcase the issues.

**Comment by danny-burrows** [comment]
I've added the functionality we discussed above, the `clear_region` and `append_lines` methods. I've also added some tests for the `insert_before` method. I hope this helps! Things do look a lot cleaner now having these tests, thanks for the suggestion! :)

**Comment by kdheepak** [comment]
Slightly related, maybe we should set a threshold in codecov.yml?

https://docs.codecov.com/docs/commit-status#threshold

**Comment by danny-burrows** [comment]
> Slightly related, maybe we should set a threshold in codecov.yml?
> 
> https://docs.codecov.com/docs/commit-status#threshold

I'd be happy to add one if you think it makes sense in this PR, do you have an acceptable value in mind? Alternatively, if this requires further discussion maybe it would be better suited to its own Issue/PR?

**Review by joshka — COMMENTED** [review]
The overarching context of my review is that I'm not as familiar with the workings of this code as you are, so I'm looking for help with the changes and tests to make it more clear that the changes are correct. They might mean that the feedback is a bit pedantic, and for that effect I apologize in advance. But I'd encourage you to think about this change with the perspective of "how would I explain these changes to someone that is reading it for the first time?" (because that's not far from the truth)

**Comment by joshka** [comment]
> I'd be happy to add one if you think it makes sense in this PR, do you have an acceptable value in mind? Alternatively, if this requires further discussion maybe it would be better suited to its own Issue/PR?

Another PR would be right for this https://github.com/ratatui-org/ratatui/issues/612

File: src/backend/test.rs [hunk 1]
```diff
@@ -179,6 +179,103 @@ impl Backend for TestBackend {
...
+                    .iter()
+                    .enumerate()
+                    .map(|(i, c)| {
+                        if i > self.buffer.index_of(self.pos.0, self.pos.1) {
+                            Cell::default()
+                        } else {
+                            c.clone()
+                        }
+                    })
+                    .collect()
```
[the remark this answers is not in the document]
Comment (danny-burrows) [reply src/backend/test.rs]: It can indeed, a good spot, I've neatened this up now using your suggestion.

File: src/backend/test.rs [hunk 2]
```diff
@@ -179,6 +179,103 @@ impl Backend for TestBackend {
...
+    }
+
+    fn append_lines(&mut self, n: u16) -> io::Result<()> {
+        let (cur_x, cur_y) = self.get_cursor()?;
+
+        let new_cursor_x = if cur_x.saturating_add(1) > self.width {
+            self.width.saturating_sub(1)
+        } else {
+            cur_x.saturating_add(1)
+        };
```
[the remark this answers is not in the document]
Comment (danny-burrows) [reply src/backend/test.rs]: Done this now, thanks!

File: src/backend/test.rs [hunk 3]
```diff
@@ -179,6 +179,103 @@ impl Backend for TestBackend {
...
+
+        let new_cursor_y = if cur_y.saturating_add(n) >= self.height {
+            for _ in 0..(cur_y
```
[the remark this answers is not in the document]
Comment (danny-burrows) [reply src/backend/test.rs]: It can indeed, I've corrected this now, thanks for the spot

File: src/backend/test.rs [hunk 4]
```diff
@@ -179,6 +179,103 @@ impl Backend for TestBackend {
...
+    fn append_lines(&mut self, n: u16) -> io::Result<()> {
+        let (cur_x, cur_y) = self.get_cursor()?;
+
+        let new_cursor_x = if cur_x.saturating_add(1) > self.width {
+            self.width.saturating_sub(1)
+        } else {
+            cur_x.saturating_add(1)
+        };
+
+        let new_cursor_y = if cur_y.saturating_add(n) >= self.height {
```
[the remark this answers is not in the document]
Comment (danny-burrows) [reply src/backend/test.rs]: I've simplified this logic as part of another comment so don't think this is needed. I think I've also made it slightly more readable with the `cursor_y_after_newlines > max_y` named variables. If you still think it's better to swap them I don't mind doing it. :)

File: src/backend/test.rs [hunk 5]
```diff
@@ -179,6 +179,103 @@ impl Backend for TestBackend {
...
+            for _ in 0..(cur_y
+                .saturating_add(n)
+                .saturating_sub(self.height)
+                .saturating_add(1))
```
[the remark this answers is not in the document]
Comment (danny-burrows) [reply src/backend/test.rs]: I've done this now, I think the result is a lot easier to read. Apologies I think I passed over this test implementation in a hurry.

File: src/backend/test.rs [hunk 6]
```diff
@@ -319,6 +416,127 @@ mod tests {
         backend.assert_buffer(&Buffer::with_lines(vec!["          "; 2]));
     }
 
+    #[test]
+    fn clear_region_all() {
+        let mut backend = TestBackend::new(10, 2);
+        backend.buffer = Buffer::with_lines(vec!["aaaaaaaaaa"; 2]);
```
[the remark this answers is not in the document]
Comment (danny-burrows) [reply src/backend/test.rs]: I've removed the `;2`'s and `;3`'s in various places. I can work on stopping the formatting from wrapping examples next

File: src/backend/test.rs [hunk 7]
```diff
@@ -319,6 +416,127 @@ mod tests {
...
+
+        backend.clear_region(ClearType::All).unwrap();
+        backend.assert_buffer(&Buffer::with_lines(vec!["          "; 2]));
+    }
+
+    #[test]
+    fn clear_region_after_cursor() {
+        let mut backend = TestBackend::new(10, 2);
+        backend.buffer = Buffer::with_lines(vec!["aaaaaaaaaa"; 2]);
```
[the remark this answers is not in the document]
Comment (danny-burrows) [reply src/backend/test.rs]: Done this now, tests seem to still pass, I think this case was caught by an earlier comment. Thanks!

File: src/backend/test.rs [hunk 8]
```diff
@@ -319,6 +416,127 @@ mod tests {
...
+
+        backend.set_cursor(0, 1).unwrap();
+        backend.clear_region(ClearType::AfterCursor).unwrap();
+        backend.assert_buffer(&Buffer::with_lines(vec!["aaaaaaaaaa", "a         "]));
+    }
+
+    #[test]
+    fn clear_region_before_cursor() {
+        let mut backend = TestBackend::new(10, 2);
+        backend.buffer = Buffer::with_lines(vec!["aaaaaaaaaa"; 2]);
```
[the remark this answers is not in the document]
Comment (danny-burrows) [reply src/backend/test.rs]: Done, thanks

File: src/backend/test.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (danny-burrows) [reply src/backend/test.rs]: I've done this now, it does add quite a few lines extra to the file but arguably is now easier to read. Thanks for the suggestion. I'll leave this one open so you can check this is acceptable.

File: src/backend/test.rs [hunk 9]
```diff
@@ -319,6 +416,127 @@ mod tests {
...
+            "cccccccccc",
+        ]));
+
+        backend.append_lines(1).unwrap();
+        assert_eq!(backend.get_cursor().unwrap(), (3, 2));
+        backend.assert_buffer(&Buffer::with_lines(vec![
+            "bbbbbbbbbb",
+            "cccccccccc",
+            "          ",
+        ]));
```
[the remark this answers is not in the document]
Comment (danny-burrows) [reply src/backend/test.rs]: I've broken these up until multiple tests, the names suggested seemed appropriate. Hopefully, this combined with some comments makes this a bit clearer now. Please do let me know if this needs further clarification.

File: src/terminal.rs [hunk 10]
```diff
@@ -471,38 +471,48 @@ where
...
-            height.saturating_sub(self.last_known_size.bottom() - self.viewport_area.top());
+
+        // Move the viewport accordingly - if it's not already at the bottom of the terminal
+        self.set_viewport_area(Rect {
+            x: self.viewport_area.left(),
+            y: (self.viewport_area.top() + height)
+                .min(self.last_known_size.bottom() - self.viewport_area.height),
+            width: self.viewport_area.width,
+            height: self.viewport_area.height,
+        });
```
Comment (joshka) [inline src/terminal.rs:484]:
```suggestion
        // Move the viewport by height, but don't move it past the bottom of the terminal
        let viewport_at_bottom = self.last_known_size.bottom() - self.viewport_area.height;
        self.set_viewport_area(Rect {
            y: self.viewport_area.y.saturating_add(height)).min(viewport_at_bottom),
            .. self.viewport
        });
```

File: src/terminal.rs [hunk 11]
```diff
@@ -471,38 +471,48 @@ where
...
+                (
+                    x,
+                    self.viewport_area.top().saturating_sub(chunk_size) + y,
+                    c,
+                )
+            });
+            self.backend.draw(iter)?;
+            self.backend.flush()?;
+            self.set_cursor(self.viewport_area.left(), self.viewport_area.top())?;
+        }
```
Comment (joshka) [inline src/terminal.rs:515]: This is fairly complex, and I'm not sure I understand it correctly. It would be helpful to explain this with a bit more of a comment. I'm going to write out an example to help clarify my understanding (I'm trying to use numbers that don't accidentally overlap to make it clearer)

Say the screen size is 40 lines, the viewport is 25 lines high starting at line 5, and we want to add 20 more lines, then I think the effect of this method is to move the viewport to start at line 15, then we create a 20 line buffer and call the draw function. Then we divide the buffer into 2 chunks sized 15 and 5.

In the first loop iteration we append (25-1+15) = 39 lines (why?) and then iterate over the chunk and draw those lines into y=0-15.

In the second loop iteration we append (25-1+5) = 29 lines and then iterate over the chunk and draw those lines into y=0-5

I wonder if `self.backend.append_lines(self.viewport_area.height.saturating_sub(1) + chunk_size)?;` is meant to be `self.backend.append_lines(self.viewport_area.height.saturating_sub(1 + chunk_size))?;`

I wonder if perhaps to test this, we should use a backend trait implementation that records each operation (i.e. we're looking at how terminal and backed interact rather than the output).

**Review by joshka — COMMENTED** [review]
Thanks for making the changes - I understand `TestBackend::append_lines()` much better now. Lots of detailed feedback again that I hope will make this better, and I think there's a couple of possible bugs that I'm not sure of.

File: src/backend/test.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (joshka) [reply src/backend/test.rs]: resolving in favor of updated comments

File: src/backend/test.rs [hunk 12]
```diff
@@ -179,6 +179,103 @@ impl Backend for TestBackend {
...
+            }
+            cur_y.saturating_add(n).min(self.height.saturating_sub(1))
+        } else {
+            cur_y.saturating_add(n)
+        };
+
+        self.set_cursor(new_cursor_x, new_cursor_y)?;
+
+        Ok(())
+    }
```
[the remark this answers is not in the document]
Comment (joshka) [reply src/backend/test.rs]: resolving in favor of updated comments

File: src/backend/test.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (joshka) [reply src/backend/test.rs]: This looks much better as the tests make it obvious what they are testing and make it easier to catch odd implementation details.

File: src/backend/test.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (joshka) [reply src/backend/test.rs]: LGTM

File: tests/terminal.rs [hunk 13]
```diff
@@ -47,3 +50,96 @@ fn terminal_draw_returns_the_completed_frame() -> Result<(), Box<dyn Error>> {
     assert_eq!(frame.area, Rect::new(0, 0, 8, 8));
     Ok(())
 }
+
+#[test]
+fn terminal_insert_before_moves_viewport() -> Result<(), Box<dyn Error>> {
```
[the remark this answers is not in the document]
Comment (danny-burrows) [reply tests/terminal.rs:54]: I've now cleaned up these tests using your suggestions. I've also added some descriptions which hopefully showcase the behaviour they are trying to ensure.

I've also added a comment to address the following in `terminal_insert_before_moves_viewport` specifically because it happened to be the first although this may be something worth adding to the docstring?

> It's not clear to me the effect of calling insert_before after rather than before the draw call - I'd suspect that most of the > time we have a buffer that ha been drawn to before and after the call?

[the remark this answers is not in the document]
Comment (danny-burrows) [reply tests/terminal.rs:54]: As it would only be the y position that is changed I thought it clearer to use the visual assertion of the buffer content, I think this is now more clear with the recent test improvements but if you still wish I can do this too.

File: src/backend/test.rs [hunk 14]
```diff
@@ -179,6 +179,59 @@ impl Backend for TestBackend {
         Ok(())
     }
 
+    fn clear_region(&mut self, clear_type: super::ClearType) -> io::Result<()> {
```
[the remark this answers is not in the document]
Comment (danny-burrows) [reply src/backend/test.rs:182]: Thanks! :)

File: src/backend/test.rs [hunk 15]
```diff
@@ -310,13 +363,249 @@ mod tests {
...
+
+        backend.set_cursor(0, 1).unwrap();
+        backend.clear_region(ClearType::BeforeCursor).unwrap();
+        backend.assert_buffer(&Buffer::with_lines(vec![
+            "          ",
+            "aaaaaaaaaa",
+            "aaaaaaaaaa",
+            "aaaaaaaaaa",
+            "aaaaaaaaaa",
+        ]));
```
[the remark this answers is not in the document]
Comment (danny-burrows) [reply src/backend/test.rs:442]: Perhaps this is a lack of clarity in the cursor position being set in the tests. In this test, the cursor is set to `(0,1)` in order to ensure the first line is cleared.

The example you've given would strike me as the correct behaviour if the cursor was instead set to `(1, 0)`. Perhaps I should change it to `(1, 1)` to get the best of both worlds?

[the remark this answers is not in the document]
Comment (joshka) [reply src/backend/test.rs:442]: Ah got it - I was reading it as if it was 1,1. The test is correct. But yeah, it's worth probably in these tests to select a point part way into the line (like 5,3) which would be easier to confirm correctness.

File: src/backend/test.rs [hunk 16]
```diff
@@ -179,6 +179,59 @@ impl Backend for TestBackend {
...
+                self.buffer.content[index..=line_end_index].fill(Cell::default());
+            }
+        }
+        Ok(())
+    }
+
+    fn append_lines(&mut self, n: u16) -> io::Result<()> {
+        let (cur_x, cur_y) = self.get_cursor()?;
+
+        let new_cursor_x = cur_x.saturating_add(1).min(self.width.saturating_sub(1));
```
[the remark this answers is not in the document]
Comment (danny-burrows) [reply src/backend/test.rs:223]: Added this now, thanks

File: src/backend/test.rs [hunk 17]
```diff
@@ -179,6 +179,59 @@ impl Backend for TestBackend {
...
+                .saturating_sub(self.height)
+                .saturating_add(1)
+                .min(max_y);
+
+            self.set_cursor(0, rotate_by)?;
+            self.clear_region(ClearType::BeforeCursor)?;
+            self.buffer
+                .content
+                .rotate_left((self.width * rotate_by).into());
+            cursor_y_after_newlines.min(max_y)
```
[the remark this answers is not in the document]
Comment (danny-burrows) [reply src/backend/test.rs]: It is indeed, this has been refactored to be a tad cleaner now

File: src/backend/test.rs [hunk 18]
```diff
@@ -179,6 +179,59 @@ impl Backend for TestBackend {
...
+            ClearType::UntilNewLine => {
+                let index = self.buffer.index_of(self.pos.0, self.pos.1);
+                let line_end_index = self.buffer.index_of(self.width - 1, self.pos.1);
+                self.buffer.content[index..=line_end_index].fill(Cell::default());
+            }
+        }
+        Ok(())
+    }
+
+    fn append_lines(&mut self, n: u16) -> io::Result<()> {
```
[the remark this answers is not in the document]
Comment (danny-burrows) [reply src/backend/test.rs:219]: This was pretty much bang on, I've added it with a small clarification that the `(at most the buffer's height)` part is specific to this test backend as there is no scrolling capabilities. I've also taken your advice and refactored the branch and I do believe it is nicer, thanks!

File: src/backend/test.rs [hunk 19]
```diff
@@ -310,13 +363,249 @@ mod tests {
...
+            "aaaaaaaaaa",
+            "bbbbbbbbbb",
+            "cccccccccc",
+            "dddddddddd",
+            "eeeeeeeeee",
+        ]));
+    }
+
+    #[test]
+    fn append_multiple_lines_past_last_line() {
```
[the remark this answers is not in the document]
Comment (danny-burrows) [reply src/backend/test.rs:599]: Added these now and got them passing as part of the above discussion, thanks for the heads up :)

File: src/backend/test.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (danny-burrows) [reply src/backend/test.rs]: I've changed this to the suggested values and also moved the `after_cursor` test to a more central point too. Hopefully this is now more readable.

File: src/backend/test.rs [hunk 20]
```diff
@@ -310,13 +363,249 @@ mod tests {
...
+        backend.buffer = Buffer::with_lines(vec![
+            "bbbbbbbbbb",
+            "cccccccccc",
+            "dddddddddd",
+            "eeeeeeeeee",
+            "          ",
+        ]);
+
+        // It also moves the cursor to the right, as is common of the behaviour of
+        // terminals in raw-mode
```
[the remark this answers is not in the document]
Comment (danny-burrows) [reply src/backend/test.rs:565]: I've added these lines as part of the new docstring, as such I'll resolve this.

File: src/terminal.rs (hunk 11 above)
Comment (danny-burrows) [reply src/terminal.rs:517]: If the screen is 40 lines, a viewport of 25 starting at line 5 (index 4) and we wish to add 20 lines:

Firstly we move the viewport by height:

1. `viewport_at_bottom` is calculated as `terminal height - viewport height` 40 - 25 = 15
2. Move the viewport `(viewport.y + add 20 lines).min(viewport_at_bottom)` (4 + 20).min(15) = 15

In this case the viewport is now at the bottom of the screen.

Next, we calculate the `max_chunk_size`, this being the most we can draw to the screen at a time assuming we cannot draw over the viewport. I did this for a couple reasons;

- We can only draw to the screen what we can see, then `append_lines` will scroll this out of the view of the terminal, so this maximum accommodates for the cursor position (which is at the start of the viewport in this point in time) and allows us to insert `n > terminal height`.
- I was hoping there would be a way to remove the `self.clear()` in order to reduce the flickering that I've been experiencing at high throughput. This is yet to bear fruit. The concept however is that if newlines are only appended *before* the viewport then maybe there is less reason to clear the viewport.

Back to the example with `viewport_area.top()` now at 15:
max_chunk_size = 15 * width (or 15 lines)

Now conceptually the for loop scrolls the terminal by 15 and then inserts 15-line chunks of the buffer at a time.

#### Alternative approaches

I see that a possible alternative is perhaps to instead use the entire terminal screen as the `max_chunk_size` but I think this precludes us to potential solutions to removing the `self.clear()`, perhaps something along the lines of using the diffing capabilities instead, in the future.

#### Comment on testing

> I wonder if perhaps to test this, we should use a backend trait implementation that records each operation (i.e. we're looking at how terminal and backed interact rather than the output).

I think this would be a great deal of work to implement on a TestBackend of sorts. I could potentially see the benefit but personally I think the tests I've added to `test/terminal.rs` cover the cases I was initially concerned with.

I've done some manual testing along simular lines, I've tested with `time::sleeps` in a couple of configurations making sure the step-by-step behaviour makes sense and it does indeed fix the issues I had (minus the flickering).

---

I hope this explanation has provided some clarity on my thoughts about the process. Thanks for your feedback :)

File: src/terminal.rs (hunk 10 above)
Comment (danny-burrows) [reply src/terminal.rs]: This is applied now, thanks!

File: tests/terminal.rs [hunk 21]
```diff
@@ -47,3 +49,147 @@ fn terminal_draw_returns_the_completed_frame() -> Result<(), Box<dyn Error>> {
...
+    })?;
+
+    assert_buffer_eq!(
+        terminal.backend().buffer().clone(),
+        Buffer::with_lines(vec![
+            "------ Line 1 ------",
+            "------ Line 2 ------",
+            "[---- Viewport ----]",
+            "                    ",
+            "                    ",
```
Comment (joshka) [inline tests/terminal.rs:90]: I really like how this makes the test super-obvious!

File: src/terminal.rs (hunk 11 above)
Comment (joshka) [reply src/terminal.rs:517]: Ok - got it. I understand the main part of this now.

I think you're right that setting up TestBackend to test this would be difficult. I think a more manual approach will have to do for testing this on the real terminal for now.

**Review by joshka — APPROVED** [review]

File: tests/terminal.rs (hunk 21 above)
Comment (danny-burrows) [reply tests/terminal.rs:90]: Thanks! But you deserve the credit for the suggestion :)

**Comment by joshka** [comment]
I've merged this - thanks for sticking with it to get it over the line. It should be available in the alpha release this Saturday.
